templates/e2e-dspy-workflow/README.ipynb

```diff
@@ -0,0 +1,1074 @@
+      "4065\n"
+     ]
+    }
+   ],
+   "source": [
+    "from dspy.teleprompt.finetune_teleprompter import bootstrap_data, convert_to_module_level_message_data\n",
+    "from src import delete_labels, NUM_THREADS, write_jsonl\n",
+    "from src.data_preprocess import valid_label_metric\n",
+    "\n",
+    "# For realism of this scenario, we are going to delete all our labels except for our test set(which is cheating and we wouldn't have in production) and our 100 true labeled examples\n",
+    "ft_trainset_to_label = delete_labels(ft_trainset)\n",
+    "\n",
+    "\n",
+    "with dspy.context(lm=llama_70b):\n",
+    "    collected_data = bootstrap_data(vanilla_program, ft_trainset_to_label, num_threads=NUM_THREADS, max_errors=10000, metric=valid_label_metric)\n",
+    "    # Make sure to only include the labels we are actively using or that arent hallucinated by the oracle\n",
+    "    collected_data_filtered = [x for x in collected_data if x[\"prediction\"][\"label\"] in labels_in_use]\n",
+    "    \n",
+    "    dataset = convert_to_module_level_message_data(collected_data_filtered, program=vanilla_program, exclude_demos=True)\n",
+    "\n",
+    "    dataset_formatted = [{\"messages\": item} for item in dataset]\n",
+    "\n",
+    "print(dataset[0])\n",
+    "print(len(dataset))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Nice utility to save the data in case you do not run the notebook all the way through\n",
+    "if True:\n",
+    "    with open(\"collected_data_filtered.jsonl\", \"w\") as f:\n",
+    "        for item in collected_data_filtered:\n",
+    "            f.write(ujson.dumps({\"example\": item[\"example\"], \"prediction\": item[\"prediction\"]}) + \"\\n\")\n",
+    "else:\n",
+    "    with open(\"collected_data_filtered.jsonl\", \"r\") as f:\n",
+    "        collected_data_filtered = [ujson.loads(line) for line in f]\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Fine-tuning\n",
+    "\n",
+    "We will use LLM Forge to fine-tune the 1B model.\n",
+    "\n",
+    "In order to do this, we need to format our data into the correct format (Follows OpenAI messaging format).\n",
+    "\n",
+    "We can let DSPy do the rest, where it will properly generate the config and run the finetuning."
```

Comment:
This can be better highlighted that DSPy comes with an `anyscale` client out of the box for fine-tuning on Anyscale?